src/Middleware/Diagnostics/src/StatusCodePage/StatusCodePagesMiddleware.cs

```diff
@@ -41,10 +42,12 @@ public async Task Invoke(HttpContext context)
     {
         var statusCodeFeature = new StatusCodePagesFeature();
         context.Features.Set<IStatusCodePagesFeature>(statusCodeFeature);
+        var endpoint = context.GetEndpoint();
```

Comment:
Having `context.GetEndpoint();` before `_next(context)` implies that `UseStatusCodePages` should be before `UseRouting` which might not be the case everywhere, including the [sample in the docs](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/error-handling?view=aspnetcore-6.0#usestatuscodepages).

This could be after `_next(context)` and also only be called if `statusCodeFeature.Enabled` is `true`.

Reply:
That's a fair consideration, and we would want to make sure the doc was accurate.

Do we have examples of anything else that's endpoint aware but placed before UseRouting?

This ordering was recommended so that the attribute behavior could be applied early and then overridden by other middleware/code while processing the request. If placed after _next, you wouldn't be able to tell if IStatusCodePagesFeature had the default value or if it was set by something in the app.

Reply:
@Tratcher Sorry, I'm confused a little bit! Are you saying that going forward `StatusCodePagesMiddleware` should always be after `RoutingMiddleware`?

Reply:
StatusCodePagesMiddleware can stay where it is and continue to work for existing apps, but if you want it to be route aware and use the SkipStatusCodePages attribute then it will need to be placed after routing.

Reply:
Oh I see! The implementation has been changed after I add the original comment. It's all clear now.

I assume if StatusCodePagesMiddleware with re-execute would be used after routing it will automatically add the routing, right?
https://github.com/dotnet/aspnetcore/blob/4656f159492b1fb021992f52caf82e5bb5a66ee7/src/Middleware/Diagnostics/src/StatusCodePage/StatusCodePagesExtensions.cs#L179-L189

What if `StatusCodePagesFeature` set the Enable property lazily? In this case it doesn't matter where `StatusCodePagesMiddleware` is as long as the feature is accessed after routing, right?
```csharp
public class StatusCodePagesFeature : IStatusCodePagesFeature
{
    private readonly HttpContext httpContext;
    public StatusCodePagesFeature(HttpContext httpContext)
    {
        _httpContext = httpContext;
    }

    private bool? _enabled;
    public bool Enabled
    {
        get 
        {
            if (_enabled == null)
            {
                var endpoint = _httpContext.GetEndpoint();
                var skipStatusCodePageMetadata = endpoint?.Metadata.GetMetadata<ISkipStatusCodePagesMetadata>();
                if (skipStatusCodePageMetadata is not null)
                {
                    _enabled = false;
                }
                else
                {
                    _enabled = true;
                }
            }
            
            return _enabled;
        }
        set
        {
            _enabled = value;
        }
    }
}
```

Reply:
StatusCodePagesFeature is a public class, so if we were to change it, we'd have to make it continue to work as it does today when called with the empty constructor. I think it might be worthwhile to just use a new internal IStatusCodePagesFeature that lazily reads the endpoint.

The only thing I don't like is locking in the wrong value for Enabled if the property is read before UseRouting(). You could definitely argue it's no worse than today where it's locked in immediately when the feature is added, but that feels a little less surprising. I wonder if it'd be better to avoid negative caching unless the property is explicitly set, but maybe that's even more surprising.